### PR TITLE
Add CRUD for student and endpoints for schools

### LINKED ISSUE #5: Create School Model and Endpoints
Screenshot from 2020-09-22 14-54-35

- think about if we need other properties in the school model (maybe country?)
- create the endpoints like in the screenshot
- add validation to class endpoint, that we require an existing school for a class to be created

### LINKED ISSUE #6: Add Students Endpoint
`/classes/:id/students`
- **GET** get list of students in the class  *--getStudents(ObjectId: classId)*
- **POST** add a new student to a class   *-- addStudent([ObjectId] : classId)*
validate whether the student already exists and create the user if it doesn’t 
send emails with login button or invitation if the student already exists

`/classes/:id/students/:id`
**DELETE** remove a student from a class _--removeStudent(ObjectId: classId)_

### PR DESCRIPTION
**Pull request by hk-zh**
endpoint: /schools and /schools/:id

### CODE REVIEW

File: src/controllers/class.controller.js [hunk 1]
```diff
@@ -27,14 +27,32 @@ const deleteClass = catchAsync(async (req, res) => {
 });
 
 const getAllClasses = catchAsync(async (req, res) => {
-  const _classes = await classService.getAllClasses(req.user._id, req.user._id);
+  const _classes = await classService.getAllClasses(req.user._id);
```
Comment (kredde) [inline src/controllers/class.controller.js:30]: why did you undo this change? we need the teacher id to just return the classes of the teacher

File: src/services/class.service.js [hunk 2]
```diff
@@ -48,6 +48,20 @@ const deleteClassById = async (classId, teacher) => {
   return _class;
 };
 
+const getStudents = async (classId) => {
+  return Class.find({ _id: classId }).students;
```
Comment (kredde) [inline src/services/class.service.js]: you probably need to populate this request in order to return not just ids but the objects

File: src/routes/v1/school.route.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ *         "403":
+ *           $ref: '#/components/responses/Forbidden'
+ *         "404":
+ *           $ref: '#/components/responses/NotFound'
+ */
+
+/**
+ * @swagger
+ * path:
+ *   /schools?name={subs}:
```
Comment (kredde) [inline src/routes/v1/school.route.js]: we dont need the parameter here in the path, its already defined in further down in parameters

File: src/routes/v1/school.route.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+ *     get:
+ *       summary: get schools
+ *       description: get schools by keyword
+ *       tags: [Schools]
+ *       security:
+ *         - bearerAuth: []
+ *       parameters:
+ *        - in: query
+ *          name: subs
```
Comment (kredde) [inline src/routes/v1/school.route.js]: name is the parameter not subs

File: src/validations/school.validation.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+const getSchool = {
+  params: Joi.object().keys({
+    schoolId: Joi.string().custom(objectId)
+  })
+};
+
+const getSchools = {
+  params: Joi.object().keys({
+    subs: Joi.string()
```
Comment (kredde) [inline src/validations/school.validation.js]: name should be the parameter

File: src/services/school.service.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const { School } = require('../models');
+
+const createSchool = async (schoolBody) => {
+  const school = await School.create(schoolBody);
+  return school;
+};
+
+const getSchools = async (subs) => {
+  return School.find({ name: { $regex: subs } });
```
Comment (kredde) [inline src/services/school.service.js]: what if someone does not pass a `subs` then this request should return all schools

File: src/controllers/class.controller.js (hunk 1 above)
Comment (hk-zh) [reply src/controllers/class.controller.js:30]: ahhh, i mean there is double times req.user._id?

File: src/routes/v1/school.route.js (hunk 3 above)
Comment (hk-zh) [reply src/routes/v1/school.route.js]: so i should just write like /schools. that's all

File: src/services/school.service.js (hunk 6 above)
Comment (hk-zh) [reply src/services/school.service.js]: so if someone input nothing, he should get nothing, right?

Comment (hk-zh) [reply src/services/school.service.js]: so how would you suggest what to write here?

File: src/controllers/class.controller.js (hunk 1 above)
Comment (kredde) [reply src/controllers/class.controller.js:30]: Oh i am sorry you are right

File: src/routes/v1/school.route.js (hunk 3 above)
Comment (kredde) [reply src/routes/v1/school.route.js]: yes here yes

File: src/services/class.service.js [hunk 7]
```diff
@@ -48,6 +48,20 @@ const deleteClassById = async (classId, teacher) => {
...
+};
+
+const addStudent = async (classId, userId) => {
+  Class.update({ _id: classId }, { $push: { students: userId } });
```
Comment (kredde) [inline src/services/class.service.js]: you are pushing the id of the teacher into the class.
Here we need to create a new Student model and add the student to the class

File: src/services/school.service.js (hunk 6 above)
Comment (hk-zh) [reply src/services/school.service.js]: `  if (!name || name.length === 0) {
    return School.find({ name: { $regex: name } });
  }`
can i write like this?

File: src/services/class.service.js (hunk 7 above)
Comment (kredde) [reply src/services/class.service.js]: you also forgot the `await` before update

File: src/services/school.service.js (hunk 6 above)
Comment (kredde) [reply src/services/school.service.js]:
```js
if (!name || name.length === 0) { 
  return School.find({} 
}
return School.find({ name: { $regex: subs } });
```

Comment (hk-zh) [reply src/services/school.service.js]: so this still means that i get all of schools when i input nothing

Comment (kredde) [reply src/services/school.service.js]: yes, maybe it already works like you have done it, you have to check

Comment (hk-zh) [reply src/services/school.service.js]: let me try

Comment (hk-zh) [reply src/services/school.service.js]: actually i can not input nothing because it is written as required.

File: src/services/class.service.js (hunk 7 above)
Comment (hk-zh) [reply src/services/class.service.js]: what if the student has already been created?

Comment (kredde) [reply src/services/class.service.js]: We have to decide to either accept only a student id here and add it (then the student must be created before), but i would prefer if we can accept an object here with all the properties of a student, then we check if the student already exists (using the email) and if he does we add the existing students id, otherwise we create a student and add the id of the newly created student

Comment (hk-zh) [reply src/services/class.service.js]: then i leave the problem here.

File: src/docs/components.yml [hunk 8]
```diff
@@ -14,11 +14,15 @@ components:
...
-        id: 5ebac534954b54139806c112
         email: fake@example.com
         firstName: Max
         lastName: Test
         role: teacher
+    Users:
+      type: array
+      items:
+        type: object
+        $ref: #/components/schemas/User'
```
Comment (kredde) [inline src/docs/components.yml]: you are missing the `'`in the beginnning here

**Review by kredde — APPROVED** [review]

**Review by kredde — APPROVED** [review]
